pkg/queue/instance.go

```diff
@@ -58,7 +65,30 @@ func (q *queueImpl) Push(item Task) {
 	q.cond.Signal()
 }
 
+// Run start the task handler, which should be in a separate go routine.
+// The queue can be closed via stop channel. However when it is being closed,
+// the left elements in the queue will be processed.
+// Note: The queue can be rerun only after closed.
 func (q *queueImpl) Run(stop <-chan struct{}) {
+	q.mutex.Lock()
+	if !q.closed {
+		q.mutex.Unlock()
+		panic("queue can not be run twice")
+	}
+	q.closed = false
+	q.mutex.Unlock()
+
+	defer func() {
+		q.mutex.Lock()
+		// set closed status
+		q.closed = true
+		q.mutex.Unlock()
+	}()
+
+	// enable rerun the queue
+	q.cond.L.Lock()
+	q.closing = false
+	q.cond.L.Unlock()
```

Comment:
I think we can reset `q.closing`  when  shutdown queue at  `if len(q.tasks) == 0 `